<?xml version="1.0" ?>
<!-- =================================================================================== -->
<!-- |    This document was autogenerated by xacro from both_psms.urdf.xacro           | -->
<!-- |    EDITING THIS FILE BY HAND IS NOT RECOMMENDED                                 | -->
<!-- =================================================================================== -->
<robot name="dvrk_psm" xmlns:xacro="http://www.ros.org/wiki/xacro">
  <!-- Macros -->
  <link name="world">
    <origin xyz="0.0 0.0 0.0"/>
  </link>
  <joint name="world_joint" type="fixed">
    <parent link="world"/>
    <child link="camera"/>
    <origin rpy="0.0 0.0 0.0" xyz="0.0 0.0 0.0"/>
  </joint>
  <link name="camera">
    <origin rpy="0.0 0.0 0.0" xyz="0.0 0.2 0.8"/>
  </link>
  <joint name="one_fixed_joint" type="fixed">
    <parent link="camera"/>
    <child link="one_remote_center_link"/>
    <origin rpy="0.026 0.204 -2.927" xyz="-0.015 -0.015 0.492"/>
  </joint>
  <link name="one_remote_center_link"/>
  <joint name="one_rc_fixed_joint" type="fixed">
    <parent link="one_remote_center_link"/>
    <child link="one_psm_base_link"/>
    <origin rpy="0 0 0" xyz="0 -0.369 0"/>
  </joint>
  <link name="one_psm_base_link">
    <visual>
      <origin rpy="1.5708 0 3.1416" xyz="0.039 -0.03888 -0.07879"/>
      <geometry>
        <mesh filename="package://dvrk_model/meshes/psm/psm_base.dae"/>
      </geometry>
    </visual>
  </link>
  <joint name="one_outer_yaw_joint" type="revolute">
    <parent link="one_psm_base_link"/>
    <child link="one_outer_yaw_link"/>
    <axis xyz="0 0 1"/>
    <origin rpy="1.5708 -1.5708 0" xyz="0.0 0.0 0.0"/>
    <limit effort="1000" lower="-1.5707" upper="1.5707" velocity=".4"/>
  </joint>
  <link name="one_outer_yaw_link">
    <visual>
      <!-- <origin rpy="0 0 -1.5708" xyz="0.0125 0 0.1575"/> -->
      <origin rpy="3.1416 0 1.5708" xyz="0.0125 0 0.1575"/>
      <geometry>
        <mesh filename="package://dvrk_model/meshes/psm/outer_yaw.dae"/>
      </geometry>
    </visual>
  </link>
  <joint name="one_outer_pitch_joint_1" type="revolute">
    <parent link="one_outer_yaw_link"/>
    <child link="one_outer_pitch_back_link"/>
    <axis xyz="0 0 1"/>
    <origin rpy="-1.5708 -1.5708 0" xyz="0 0.0295 0.1495"/>
    <limit effort="1000" lower="-0.7854" upper="0.7854" velocity=".4"/>
  </joint>
  <link name="one_outer_pitch_back_link">
    <visual>
      <origin rpy="0 0 -0.27129" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://dvrk_model/meshes/psm/outer_pitch_back.dae"/>
      </geometry>
    </visual>
  </link>
  <joint name="one_outer_pitch_joint_2" type="continuous">
    <parent link="one_outer_yaw_link"/>
    <child link="one_outer_pitch_front_link"/>
    <axis xyz="0 0 1"/>
    <origin rpy="-1.5708 -1.5708 0" xyz="0 0.0295 0.0595"/>
    <mimic joint="one_outer_pitch_joint_1" multiplier="1"/>
  </joint>
  <link name="one_outer_pitch_front_link">
    <visual>
      <origin rpy="0 0 -0.27129" xyz="0 0 0"/>
      <!-- <origin rpy="0 0 0" xyz="0 0 0"/> -->
      <geometry>
        <mesh filename="package://dvrk_model/meshes/psm/outer_pitch_front.dae"/>
      </geometry>
    </visual>
  </link>
  <joint name="one_outer_pitch_joint_3" type="continuous">
    <parent link="one_outer_pitch_back_link"/>
    <child link="one_outer_pitch_bottom_link"/>
    <axis xyz="0 0 1"/>
    <origin rpy="0 0 0" xyz="0.04178 0.15007 -0.0137"/>
    <mimic joint="one_outer_pitch_joint_1" multiplier="-1"/>
  </joint>
  <link name="one_outer_pitch_bottom_link">
    <visual>
      <origin rpy="0 -1.5708 0" xyz="0.009 0 0"/>
      <geometry>
        <mesh filename="package://dvrk_model/meshes/psm/outer_pitch_bottom.dae"/>
      </geometry>
    </visual>
  </link>
  <joint name="one_outer_pitch_joint_4" type="continuous">
    <parent link="one_outer_pitch_back_link"/>
    <child link="one_outer_pitch_top_link"/>
    <axis xyz="0 0 1"/>
    <origin rpy="0 0 0" xyz="0.04209 0.18695 -0.02412"/>
    <mimic joint="one_outer_pitch_joint_1" multiplier="-1"/>
  </joint>
  <link name="one_outer_pitch_top_link">
    <visual>
      <origin rpy="0 -1.5708 0" xyz="0.009 0 0"/>
      <geometry>
        <mesh filename="package://dvrk_model/meshes/psm/outer_pitch_top.dae"/>
      </geometry>
    </visual>
  </link>
  <joint name="one_outer_pitch_joint_5" type="continuous">
    <parent link="one_outer_pitch_bottom_link"/>
    <child link="one_outer_insertion_link"/>
    <axis xyz="0 0 1"/>
    <origin rpy="0 0 0" xyz="-0.520 0 -0.0155"/>
    <mimic joint="one_outer_pitch_joint_1" multiplier="1"/>
  </joint>
  <link name="one_outer_insertion_link">
    <visual>
      <origin rpy="0 -1.5708 1.5708" xyz="0.02528 0.429 0"/>
      <geometry>
        <mesh filename="package://dvrk_model/meshes/psm/outer_insertion.dae"/>
      </geometry>
    </visual>
  </link>
  <joint name="one_outer_insertion_joint" type="prismatic">
    <parent link="one_outer_insertion_link"/>
    <child link="one_tool_main_link"/>
    <axis xyz="0 0 1"/>
    <origin rpy="1.5708 0 0" xyz="-0.0403 0.23622 0"/>
    <limit effort="1000" lower="0" upper="0.240" velocity=".4"/>
  </joint>
  <link name="one_tool_main_link">
    <visual>
      <origin rpy="0 0 1.5708" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://dvrk_model/meshes/psm/tool_main.dae"/>
      </geometry>
    </visual>
  </link>
  <joint name="one_outer_roll_joint" type="revolute">
    <parent link="one_tool_main_link"/>
    <child link="one_tool_wrist_link"/>
    <axis xyz="0 0 1"/>
    <origin rpy="0 0 0" xyz="0 0 0.36680"/>
    <limit effort="1000" lower="-2.2689" upper="2.2689" velocity=".4"/>
  </joint>
  <link name="one_tool_wrist_link">
    <visual>
      <origin rpy="0 0 1.5708" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://dvrk_model/meshes/psm/tool_wrist_link.dae"/>
      </geometry>
    </visual>
  </link>
  <joint name="one_outer_roll_shaft_joint" type="fixed">
    <parent link="one_tool_wrist_link"/>
    <child link="one_tool_wrist_shaft_link"/>
    <axis xyz="0 0 1"/>
    <origin rpy="0 0 0" xyz="0 0 0.0091"/>
  </joint>
  <link name="one_tool_wrist_shaft_link">
    <visual>
      <origin rpy="1.5708 0 0" xyz="0 0.00401 0"/>
      <geometry>
        <mesh filename="package://dvrk_model/meshes/psm/tool_wrist_shaft_link.dae"/>
      </geometry>
    </visual>
  </link>
  <joint name="one_outer_wrist_pitch_joint" type="revolute">
    <parent link="one_tool_wrist_shaft_link"/>
    <child link="one_tool_wrist_sca_link"/>
    <axis xyz="0 0 1"/>
    <origin rpy="-1.5708 -1.5708 0" xyz="0 0 0"/>
    <limit effort="1000" lower="-1.5707" upper="1.5707" velocity=".4"/>
  </joint>
  <link name="one_tool_wrist_sca_link">
    <visual>
      <!-- <origin rpy="0 3.1416 0" xyz="0 0 0"/> -->
      <origin rpy="1.5708 3.1416 0" xyz="0.0051 0.0032 0"/>
      <geometry>
        <mesh filename="package://dvrk_model/meshes/psm/tool_wrist_sca_link.dae"/>
      </geometry>
    </visual>
  </link>
  <joint name="one_outer_wrist_yaw_joint" type="revolute">
    <parent link="one_tool_wrist_sca_link"/>
    <child link="one_tool_wrist_sca_shaft_link"/>
    <axis xyz="0 0 1"/>
    <origin rpy="-1.5708 -1.5708 0" xyz="0.0089 0 0"/>
    <!-- +/- 80 deg -->
    <limit effort="1000" lower="-1.3963" upper="1.3963" velocity=".4"/>
  </joint>
  <link name="one_tool_wrist_sca_shaft_link">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 -0.0025"/>
      <geometry>
        <mesh filename="package://dvrk_model/meshes/psm/tool_wrist_sca_shaft_link.dae"/>
      </geometry>
    </visual>
  </link>
  <joint name="one_outer_wrist_open_angle_joint_1" type="revolute">
    <parent link="one_tool_wrist_sca_shaft_link"/>
    <child link="one_tool_wrist_sca_ee_link_1"/>
    <axis xyz="0 0 1"/>
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <!-- <origin rpy="-1.5708 -1.5708 0" xyz="0 0 0"/> -->
    <limit effort="1000" lower="0" upper="1.5707" velocity=".4"/>
  </joint>
  <link name="one_tool_wrist_sca_ee_link_1">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://dvrk_model/meshes/psm/tool_wrist_sca_link_2.dae"/>
      </geometry>
    </visual>
  </link>
  <joint name="one_outer_wrist_open_angle_joint_2" type="revolute">
    <parent link="one_tool_wrist_sca_shaft_link"/>
    <child link="one_tool_wrist_sca_ee_link_2"/>
    <axis xyz="0 0 1"/>
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <limit effort="1000" lower="-1.5707" upper="1.5707" velocity=".4"/>
    <!--<mimic joint="${prefix}outer_wrist_open_angle_joint_1" multiplier="-1"/> -->
  </joint>
  <link name="one_tool_wrist_sca_ee_link_2">
    <visual>
      <origin rpy="0 3.1516 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://dvrk_model/meshes/psm/tool_wrist_sca_link_2.dae"/>
      </geometry>
    </visual>
  </link>
  <joint name="one_fake_joint_1" type="fixed">
    <limit effort="1000.0" lower="0.0" upper="3.1415" velocity="0.5"/>
    <origin rpy="0 0 1.5708" xyz="0 0 0"/>
    <parent link="one_tool_wrist_sca_shaft_link"/>
    <child link="one_tool_base_link"/>
  </joint>
  <link name="one_tool_base_link">
  </link>
  <joint name="one_fake_joint_2" type="fixed">
    <limit effort="1000.0" lower="0.0" upper="3.1415" velocity="0.5"/>
    <origin rpy="0 0 0" xyz="0.00483 0 0"/>
    <parent link="one_tool_base_link"/>
    <child link="one_tool_middle_link"/>
  </joint>
  <link name="one_tool_middle_link">
  </link>
  <joint name="one_fake_joint_3" type="fixed">
    <limit effort="1000.0" lower="0.0" upper="3.1415" velocity="0.5"/>
    <origin rpy="0 0 0" xyz="0.00483 0 0"/>
    <parent link="one_tool_middle_link"/>
    <child link="one_tool_tip_link"/>
  </joint>
  <link name="one_tool_tip_link">
  </link>
  <joint name="two_fixed_joint" type="fixed">
    <parent link="camera"/>
    <child link="two_remote_center_link"/>
    <origin rpy="0.0 0.0 3.1415926" xyz="0.25 0.0 0.5"/>
  </joint>
  <link name="two_remote_center_link"/>
  <joint name="two_rc_fixed_joint" type="fixed">
    <parent link="two_remote_center_link"/>
    <child link="two_psm_base_link"/>
    <origin rpy="0 0 0" xyz="0 -0.369 0"/>
  </joint>
  <link name="two_psm_base_link">
    <visual>
      <origin rpy="1.5708 0 3.1416" xyz="0.039 -0.03888 -0.07879"/>
      <geometry>
        <mesh filename="package://dvrk_model/meshes/psm/psm_base.dae"/>
      </geometry>
    </visual>
  </link>
  <joint name="two_outer_yaw_joint" type="revolute">
    <parent link="two_psm_base_link"/>
    <child link="two_outer_yaw_link"/>
    <axis xyz="0 0 1"/>
    <origin rpy="1.5708 -1.5708 0" xyz="0.0 0.0 0.0"/>
    <limit effort="1000" lower="-1.5707" upper="1.5707" velocity=".4"/>
  </joint>
  <link name="two_outer_yaw_link">
    <visual>
      <!-- <origin rpy="0 0 -1.5708" xyz="0.0125 0 0.1575"/> -->
      <origin rpy="3.1416 0 1.5708" xyz="0.0125 0 0.1575"/>
      <geometry>
        <mesh filename="package://dvrk_model/meshes/psm/outer_yaw.dae"/>
      </geometry>
    </visual>
  </link>
  <joint name="two_outer_pitch_joint_1" type="revolute">
    <parent link="two_outer_yaw_link"/>
    <child link="two_outer_pitch_back_link"/>
    <axis xyz="0 0 1"/>
    <origin rpy="-1.5708 -1.5708 0" xyz="0 0.0295 0.1495"/>
    <limit effort="1000" lower="-0.7854" upper="0.7854" velocity=".4"/>
  </joint>
  <link name="two_outer_pitch_back_link">
    <visual>
      <origin rpy="0 0 -0.27129" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://dvrk_model/meshes/psm/outer_pitch_back.dae"/>
      </geometry>
    </visual>
  </link>
  <joint name="two_outer_pitch_joint_2" type="continuous">
    <parent link="two_outer_yaw_link"/>
    <child link="two_outer_pitch_front_link"/>
    <axis xyz="0 0 1"/>
    <origin rpy="-1.5708 -1.5708 0" xyz="0 0.0295 0.0595"/>
    <mimic joint="two_outer_pitch_joint_1" multiplier="1"/>
  </joint>
  <link name="two_outer_pitch_front_link">
    <visual>
      <origin rpy="0 0 -0.27129" xyz="0 0 0"/>
      <!-- <origin rpy="0 0 0" xyz="0 0 0"/> -->
      <geometry>
        <mesh filename="package://dvrk_model/meshes/psm/outer_pitch_front.dae"/>
      </geometry>
    </visual>
  </link>
  <joint name="two_outer_pitch_joint_3" type="continuous">
    <parent link="two_outer_pitch_back_link"/>
    <child link="two_outer_pitch_bottom_link"/>
    <axis xyz="0 0 1"/>
    <origin rpy="0 0 0" xyz="0.04178 0.15007 -0.0137"/>
    <mimic joint="two_outer_pitch_joint_1" multiplier="-1"/>
  </joint>
  <link name="two_outer_pitch_bottom_link">
    <visual>
      <origin rpy="0 -1.5708 0" xyz="0.009 0 0"/>
      <geometry>
        <mesh filename="package://dvrk_model/meshes/psm/outer_pitch_bottom.dae"/>
      </geometry>
    </visual>
  </link>
  <joint name="two_outer_pitch_joint_4" type="continuous">
    <parent link="two_outer_pitch_back_link"/>
    <child link="two_outer_pitch_top_link"/>
    <axis xyz="0 0 1"/>
    <origin rpy="0 0 0" xyz="0.04209 0.18695 -0.02412"/>
    <mimic joint="two_outer_pitch_joint_1" multiplier="-1"/>
  </joint>
  <link name="two_outer_pitch_top_link">
    <visual>
      <origin rpy="0 -1.5708 0" xyz="0.009 0 0"/>
      <geometry>
        <mesh filename="package://dvrk_model/meshes/psm/outer_pitch_top.dae"/>
      </geometry>
    </visual>
  </link>
  <joint name="two_outer_pitch_joint_5" type="continuous">
    <parent link="two_outer_pitch_bottom_link"/>
    <child link="two_outer_insertion_link"/>
    <axis xyz="0 0 1"/>
    <origin rpy="0 0 0" xyz="-0.520 0 -0.0155"/>
    <mimic joint="two_outer_pitch_joint_1" multiplier="1"/>
  </joint>
  <link name="two_outer_insertion_link">
    <visual>
      <origin rpy="0 -1.5708 1.5708" xyz="0.02528 0.429 0"/>
      <geometry>
        <mesh filename="package://dvrk_model/meshes/psm/outer_insertion.dae"/>
      </geometry>
    </visual>
  </link>
  <joint name="two_outer_insertion_joint" type="prismatic">
    <parent link="two_outer_insertion_link"/>
    <child link="two_tool_main_link"/>
    <axis xyz="0 0 1"/>
    <origin rpy="1.5708 0 0" xyz="-0.0403 0.23622 0"/>
    <limit effort="1000" lower="0" upper="0.240" velocity=".4"/>
  </joint>
  <link name="two_tool_main_link">
    <visual>
      <origin rpy="0 0 1.5708" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://dvrk_model/meshes/psm/tool_main.dae"/>
      </geometry>
    </visual>
  </link>
  <joint name="two_outer_roll_joint" type="revolute">
    <parent link="two_tool_main_link"/>
    <child link="two_tool_wrist_link"/>
    <axis xyz="0 0 1"/>
    <origin rpy="0 0 0" xyz="0 0 0.36680"/>
    <limit effort="1000" lower="-2.2689" upper="2.2689" velocity=".4"/>
  </joint>
  <link name="two_tool_wrist_link">
    <visual>
      <origin rpy="0 0 1.5708" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://dvrk_model/meshes/psm/tool_wrist_link.dae"/>
      </geometry>
    </visual>
  </link>
  <joint name="two_outer_roll_shaft_joint" type="fixed">
    <parent link="two_tool_wrist_link"/>
    <child link="two_tool_wrist_shaft_link"/>
    <axis xyz="0 0 1"/>
    <origin rpy="0 0 0" xyz="0 0 0.0091"/>
  </joint>
  <link name="two_tool_wrist_shaft_link">
    <visual>
      <origin rpy="1.5708 0 0" xyz="0 0.00401 0"/>
      <geometry>
        <mesh filename="package://dvrk_model/meshes/psm/tool_wrist_shaft_link.dae"/>
      </geometry>
    </visual>
  </link>
  <joint name="two_outer_wrist_pitch_joint" type="revolute">
    <parent link="two_tool_wrist_shaft_link"/>
    <child link="two_tool_wrist_sca_link"/>
    <axis xyz="0 0 1"/>
    <origin rpy="-1.5708 -1.5708 0" xyz="0 0 0"/>
    <limit effort="1000" lower="-1.5707" upper="1.5707" velocity=".4"/>
  </joint>
  <link name="two_tool_wrist_sca_link">
    <visual>
      <!-- <origin rpy="0 3.1416 0" xyz="0 0 0"/> -->
      <origin rpy="1.5708 3.1416 0" xyz="0.0051 0.0032 0"/>
      <geometry>
        <mesh filename="package://dvrk_model/meshes/psm/tool_wrist_sca_link.dae"/>
      </geometry>
    </visual>
  </link>
  <joint name="two_outer_wrist_yaw_joint" type="revolute">
    <parent link="two_tool_wrist_sca_link"/>
    <child link="two_tool_wrist_sca_shaft_link"/>
    <axis xyz="0 0 1"/>
    <origin rpy="-1.5708 -1.5708 0" xyz="0.0089 0 0"/>
    <!-- +/- 80 deg -->
    <limit effort="1000" lower="-1.3963" upper="1.3963" velocity=".4"/>
  </joint>
  <link name="two_tool_wrist_sca_shaft_link">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 -0.0025"/>
      <geometry>
        <mesh filename="package://dvrk_model/meshes/psm/tool_wrist_sca_shaft_link.dae"/>
      </geometry>
    </visual>
  </link>
  <joint name="two_outer_wrist_open_angle_joint_1" type="revolute">
    <parent link="two_tool_wrist_sca_shaft_link"/>
    <child link="two_tool_wrist_sca_ee_link_1"/>
    <axis xyz="0 0 1"/>
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <!-- <origin rpy="-1.5708 -1.5708 0" xyz="0 0 0"/> -->
    <limit effort="1000" lower="0" upper="1.5707" velocity=".4"/>
  </joint>
  <link name="two_tool_wrist_sca_ee_link_1">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://dvrk_model/meshes/psm/tool_wrist_sca_link_2.dae"/>
      </geometry>
    </visual>
  </link>
  <joint name="two_outer_wrist_open_angle_joint_2" type="revolute">
    <parent link="two_tool_wrist_sca_shaft_link"/>
    <child link="two_tool_wrist_sca_ee_link_2"/>
    <axis xyz="0 0 1"/>
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <limit effort="1000" lower="-1.5707" upper="1.5707" velocity=".4"/>
    <!--<mimic joint="${prefix}outer_wrist_open_angle_joint_1" multiplier="-1"/> -->
  </joint>
  <link name="two_tool_wrist_sca_ee_link_2">
    <visual>
      <origin rpy="0 3.1516 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://dvrk_model/meshes/psm/tool_wrist_sca_link_2.dae"/>
      </geometry>
    </visual>
  </link>
  <joint name="two_fake_joint_1" type="fixed">
    <limit effort="1000.0" lower="0.0" upper="3.1415" velocity="0.5"/>
    <origin rpy="0 0 1.5708" xyz="0 0 0"/>
    <parent link="two_tool_wrist_sca_shaft_link"/>
    <child link="two_tool_base_link"/>
  </joint>
  <link name="two_tool_base_link">
  </link>
  <joint name="two_fake_joint_2" type="fixed">
    <limit effort="1000.0" lower="0.0" upper="3.1415" velocity="0.5"/>
    <origin rpy="0 0 0" xyz="0.00483 0 0"/>
    <parent link="two_tool_base_link"/>
    <child link="two_tool_middle_link"/>
  </joint>
  <link name="two_tool_middle_link">
  </link>
  <joint name="two_fake_joint_3" type="fixed">
    <limit effort="1000.0" lower="0.0" upper="3.1415" velocity="0.5"/>
    <origin rpy="0 0 0" xyz="0.00483 0 0"/>
    <parent link="two_tool_middle_link"/>
    <child link="two_tool_tip_link"/>
  </joint>
  <link name="two_tool_tip_link">
  </link>
</robot>

